### ISSUE TITLE
How to implement custom loss functions without label assignments (unsupervised) ?

### ISSUE THREAD

**Issue by simonmaurer** [apache/incubator-mxnet#10641]
@pengwangucla @saicoco that example #5580 helped me pretty well starting to understand the data flow. thanks a lot
now I wanna implement three custom loss functions which not only have an additional parameter (specifically a hyperparameter and not learned) but also are independant of the label (as the training is unsupervised and from that new layer perspective only depends on a binary transformation of the preceding layer).
W.r.t. this I got 3 issues:
1. am I right to assume that we can either use mx.operator.CustomOp or mx.operator.NDArrayOp ?

2. if I compare MXNets implementation of forward()/backward() pass to Caffe:
`template <typename Dtype>
void CustomLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
                                  const vector<Blob<Dtype>*>& top)`
then
`def forward(self, is_train, req, in_data, out_data, aux):`
contains the data of one batch (each row corresponds to one sample) and not just one sample, correct ?

3. I like to train the network in Python and do the inference in C++ later: what is the way to do this ?
using
`c_predict_api.h,` or
`mxnet-cpp/MxNetCpp.h` as in [https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/feature_extract/feature_extract.cpp](url)
in other words will these loss layers be registered for C++ as well or do we strictly have to rely on NNVM? 

4. I still don't quite get the example on [https://mxnet.incubator.apache.org/faq/new_op.html](url). If we look at the mathematical definition of the softmax function and its layer, it's a function of the learned weight parameters _w<sub>i</sub>_ and the inputs _x<sub>i</sub>_ , so why are there no weights in the forward() pass ?

**Comment by spidyDev** [comment]
@simonmaurer  Please close the issue, if your question has been answered.

**Comment by simonmaurer** [comment]
@ThomasDelteil, @spidyDev  discussion of questions continued as stated with your link - closing the issue
